Perf optimization for conv and gemm kernels.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37829 Some TODO fixes.
* **#37626 Perf optimization for conv and gemm kernels.**
* #37624 Added per channel separate test cases for fc and deconv tests.
* #37623 Changes to enable per channel support on dynamic linear.
* #37622 Enabled per channel quantized static linear/conv
* #37621 Added per channel kernels for depthwise conv.
* #37620 Changes to enable per channel requant.
* #37619 Enable per channel zero point.
* #37618 Interface changes to enable per channel quant.

Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Benchmarking was done via q8gemm-bench and convolution-bench on pixel-2 phone.
For aarch32:
- Before this change q8gemm-bench was 1.7% slower overall. After this it was .9%.
- convolution-bench was too noisy with wide variance so did not account for it.

For aarch64 also it was inconclusive.

Generally it was hard to get a good pulse at op level benchmarking though. I left the changes in largely because they seemed like a better schedule. In out-of-order processor it may not mean much though.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

kimishpatel committed:
Perf optimization for conv and gemm kernels.

Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads.
Particularly at the start of the kernel we have latency bound chains.
It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if
this actually helps with perf espcially in OOO CPUs, but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

dreiss reviewed (changes requested):
Can you update your test plan to indicate what benchmarks you ran and what the results were.

kimishpatel commented:
> Can you update your test plan to indicate what benchmarks you ran and what the results were.

@dreiss Updated comment.

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Benchmarking was done via q8gemm-bench and convolution-bench on pixel-2 phone.
For aarch32:
- Before this change q8gemm-bench was 1.7% slower overall. After this it was .9%.
- convolution-bench was too noisy with wide variance so did not account for it.

For aarch64 also it was inconclusive.

Generally it was hard to get a good pulse at op level benchmarking though. I left the changes in largely because they seemed like a better schedule. In out-of-order processor it may not mean much though.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Benchmarking was done via q8gemm-bench and convolution-bench on pixel-2 phone.
For aarch32:
- Before this change q8gemm-bench was 1.7% slower overall. After this it was .9%.
- convolution-bench was too noisy with wide variance so did not account for it.

For aarch64 also it was inconclusive.

Generally it was hard to get a good pulse at op level benchmarking though. I left the changes in largely because they seemed like a better schedule. In out-of-order processor it may not mean much though.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S:
```diff
@@ -78,18 +78,30 @@ BEGIN_FUNCTION pytorch_q8gemm_ukernel_4x8__aarch32_neon
     # Load output channel index
-    LDR r5, [sp, 100]
+    LDR r5, [sp, 104]
+    CMP r0, 2
```
Where are you using the result of this comparison?  I don't seem to be able to spot it.

kimishpatel replied:
This is from the original code. It is just movement of code that seem to have hidden its usage. It here: https://github.com/pytorch/pytorch/pull/37626/files/20a19b900e54e205185d5ce54a3ec327e54e12c2#diff-f682e6802902a89ec92f72efe5ec2a5bR120

AshkanAliabadi replied:
This is very dangerous.  Condition flags can be implicitly set as a side effect by quite a few instructions which makes putting too much distance between condition checks and their actual usage dangerous.  For instance if someone a month from now modifies this code to add a SUBS in between this  CMP and CSEL this code will break.

kimishpatel replied:
OK, so I think you could also say that if someone is changing this code then they probably understand what they are doing. I do understand if this code gets larger it become intractable but probably you wont write really long handwritten assembly. Having said all that, in this instance I dont think there is much justification to move the cmp around. I was trying to hide latency but I doubt it helps much. I did not really account the point you mentioned, so thats fair. I will move it back to where it was.

AshkanAliabadi replied:
Again, as mentioned before, all my comments are merely suggestions and you are more than welcome to do as you see fit.

kimishpatel replied:
Made the changes.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S:
```diff
@@ -42,33 +42,37 @@ BEGIN_FUNCTION pytorch_q8gemm_ukernel_8x8__aarch64_neon
+    LDP x16, x8, [sp]
+
+    # a1
+    CMP x0, 2
```
Same.  I would have expected a condition code on ADD if that was supposed to be conditional?

kimishpatel replied:
Same as earlier. Just hidden away. https://github.com/pytorch/pytorch/pull/37626/files/20a19b900e54e205185d5ce54a3ec327e54e12c2#diff-f682e6802902a89ec92f72efe5ec2a5bR120

AshkanAliabadi replied:
Likewise.  It's should be fine to move the ADD as far as I can tell, but moving the CMP is a hard to find bug in the making. :)

AshkanAliabadi reviewed (commented):
Thanks. LGTM except for the two questions I have.

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Benchmarking was done via q8gemm-bench and convolution-bench on pixel-2 phone.
For aarch32:
- Before this change q8gemm-bench was 1.7% slower overall. After this it was .9%.
- convolution-bench was too noisy with wide variance so did not account for it.

For aarch64 also it was inconclusive.

Generally it was hard to get a good pulse at op level benchmarking though. I left the changes in largely because they seemed like a better schedule. In out-of-order processor it may not mean much though.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Benchmarking was done via q8gemm-bench and convolution-bench on pixel-2 phone.
For aarch32:
- Before this change q8gemm-bench was 1.7% slower overall. After this it was .9%.
- convolution-bench was too noisy with wide variance so did not account for it.

For aarch64 also it was inconclusive.

Generally it was hard to get a good pulse at op level benchmarking though. I left the changes in largely because they seemed like a better schedule. In out-of-order processor it may not mean much though.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Benchmarking was done via q8gemm-bench and convolution-bench on pixel-2 phone.
For aarch32:
- Before this change q8gemm-bench was 1.7% slower overall. After this it was .9%.
- convolution-bench was too noisy with wide variance so did not account for it.

For aarch64 also it was inconclusive.

Generally it was hard to get a good pulse at op level benchmarking though. I left the changes in largely because they seemed like a better schedule. In out-of-order processor it may not mean much though.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]

kimishpatel committed:
Update on "Perf optimization for conv and gemm kernels."


Summary:
Did some rescheduling of the instructions to hide latency of the loads. Particularly at the start of the kernel we have latency bound chains. It seems to improve perf form aarch32.
Also did some inst rescheduling for aarch64 gemm kernel. Not clear if this actually helps with perf especially in OOO CPUs (comment also applies to aarch32 kernels), but worth a try.

Test Plan:
qnnpack tests
q8gemm-test

Benchmarking was done via q8gemm-bench and convolution-bench on pixel-2 phone.
For aarch32:
- Before this change q8gemm-bench was 1.7% slower overall. After this it was .9%.
- convolution-bench was too noisy with wide variance so did not account for it.

For aarch64 also it was inconclusive.

Generally it was hard to get a good pulse at op level benchmarking though. I left the changes in largely because they seemed like a better schedule. In out-of-order processor it may not mean much though.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339037](https://our.internmc.facebook.com/intern/diff/D21339037)

[ghstack-poisoned]
